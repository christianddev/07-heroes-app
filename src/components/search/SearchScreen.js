import React, { useMemo } from "react";
import { useLocation } from "react-router-dom";
import queryString from "query-string";
// import { heroes } from '../../data/heroes';
import { useForm } from "../../hook/useForm";
import HeroCard from "../heroes/HeroCard";
import { getHeroesBySuperhero } from "../../selectors/getHeroByIdSuperhero";
// import { heroes } from '../../data/heroes';

const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { queryName = "" } = queryString.parse(location.search);

  const [{ name }, handleInputChange] = useForm({ name: queryName });

  const heroesFiltered = useMemo(
    () => getHeroesBySuperhero(queryName),
    [queryName]
  );
  const handleSearch = (e) => {
    e.preventDefault();
    history.push(`?queryName=${name}`);
  };

  return (
    <div>
      <h1>SearchScreen</h1>
      <hr />
      <div className="row">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={handleSearch}>
            <input
              type="text"
              placeholder="Name"
              name="name"
              value={name}
              className="form-control"
              autoComplete="off"
              onChange={handleInputChange}
            />
            <button type="submit" className="btn btn-primary m-1">
              Search
            </button>
          </form>
        </div>
        <div className="col-7  animate__animated animate__fadeInLeft">
          <h4>Results:</h4>
          <hr />
          {
            queryName === "" && 
            <div className="alert alert-info">Search a hero</div>
          }
          {
            queryName !== "" && heroesFiltered.length === 0 &&
            <div className="alert alert-danger">THre is no a hero with</div>
          }
          {
            heroesFiltered.map((hero) => (
                <HeroCard key={hero.id} hero={hero} />
              )
            )
          }
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
