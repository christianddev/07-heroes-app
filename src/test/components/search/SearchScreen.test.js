import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom';
import SearchScreen from '../../../components/search/SearchScreen';
import { historyMock } from '../../fixtures/historyMock';

describe('SearchScreen.js', () => {

  test('should default valeus', () => {
    const wrapper = mount(
      <MemoryRouter initialEntries={['/search']}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero');
  })

  test('should search some hero (batman) using queryString', () => {
    const name = 'batman';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?queryName=${name}`]}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(name);
    expect(wrapper.find('.card-body').exists()).toBe(true);
    expect(wrapper.find('.alert-danger').exists()).toBe(false);
  })
  
  test('should error if dont found hero', () => {
    const name = 'batmannnn';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?queryName=${name}`]}>
        <Route path="/search" component={SearchScreen}/>
      </MemoryRouter>
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('input').prop('value')).toBe(name);
    expect(wrapper.find('.card-body').exists()).toBe(false);
    expect(wrapper.find('.alert-danger').exists()).toBe(true);
  })

  test('should call history.push', () => {
    const name = 'batmannnn';
    const wrapper = mount(
      <MemoryRouter initialEntries={[`/search?queryName=${name}`]}>
        <Route path="/search" component={() => <SearchScreen history={historyMock} />}/>
      </MemoryRouter>
    );
    const inputTarget = {
      target: {
        name: 'name', 
        value: name
      } 
    };
    wrapper.find('input').simulate('change', inputTarget );

    wrapper.find('form').prop('onSubmit')({
      preventDefault(){}
    });

    expect(historyMock.push).toHaveBeenCalledWith(`?queryName=${name}`);
  })

})
