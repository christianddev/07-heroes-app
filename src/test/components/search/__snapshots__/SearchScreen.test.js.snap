// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchScreen.js should default valeus 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value=""
        />
        <button
          className="btn btn-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7  animate__animated animate__fadeInLeft"
    >
      <h4>
        Results:
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchScreen.js should error if dont found hero 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value="batmannnn"
        />
        <button
          className="btn btn-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7  animate__animated animate__fadeInLeft"
    >
      <h4>
        Results:
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        THre is no a hero with
      </div>
    </div>
  </div>
</div>
`;

exports[`SearchScreen.js should search some hero (batman) using queryString 1`] = `
<div>
  <h1>
    SearchScreen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-primary m-1"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7  animate__animated animate__fadeInLeft"
    >
      <h4>
        Results:
      </h4>
      <hr />
      <div
        className="card"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="com-md-4"
          >
            <img
              alt="dc-batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                Ver m√°s...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
